drivers/mspi/mspi_stm32_xspi.c

```diff
@@ -0,0 +1,1873 @@
+}
+
+/* Sends the command to configure the device according to the DTS */
+static int mspi_stm32_xspi_config_mem(const struct device *dev, uint8_t cfg_mode, uint8_t cfg_rate)
+{
+	struct mspi_stm32_data *dev_data = dev->data;
+	uint8_t reg[2];
+
+	/* MSPI_IO_MODE_SINGLE/MSPI_DATA_RATE_SINGLE is already done */
+	if ((cfg_mode == MSPI_IO_MODE_SINGLE) && (cfg_rate == MSPI_DATA_RATE_SINGLE)) {
+		return 0;
+	}
+
+	/* Write Configuration register 2 (with new dummy cycles) */
+	if (mspi_stm32_xspi_write_cfg2reg_dummy(dev, MSPI_IO_MODE_SINGLE, MSPI_DATA_RATE_SINGLE) !=
+	    0) {
+		LOG_ERR("XSPI write CFGR2 failed");
+		return -EIO;
+	}
+	if (mspi_stm32_xspi_mem_ready(dev, MSPI_IO_MODE_SINGLE, MSPI_DATA_RATE_SINGLE) != 0) {
+		LOG_ERR("XSPI autopolling failed");
+		return -EIO;
+	}
+	if (mspi_stm32_xspi_write_enable(dev, MSPI_IO_MODE_SINGLE, MSPI_DATA_RATE_SINGLE) != 0) {
+		LOG_ERR("XSPI write Enable 2 failed");
+		return -EIO;
+	}
+
+	/* Write Configuration register 2 (with Octal I/O SPI protocol : choose STR or DTR) */
+	uint8_t mode_enable = ((cfg_rate == MSPI_DATA_RATE_DUAL) ? SPI_NOR_CR2_DTR_OPI_EN
+								 : SPI_NOR_CR2_STR_OPI_EN);
+	if (mspi_stm32_xspi_write_cfg2reg_io(dev, MSPI_IO_MODE_SINGLE, MSPI_DATA_RATE_SINGLE,
+					     mode_enable) != 0) {
+		LOG_ERR("XSPI write CFGR2 failed");
+		return -EIO;
+	}
+
+	/* Wait till the configuration is effective and check that memory is ready */
+	k_busy_wait(MSPI_STM32_WRITE_REG_MAX_TIME * USEC_PER_MSEC);
+
+	/* Reconfigure the memory type of the peripheral */
+	dev_data->hmspi.xspi.Init.MemoryType = HAL_XSPI_MEMTYPE_MACRONIX;
+	dev_data->hmspi.xspi.Init.DelayHoldQuarterCycle = HAL_XSPI_DHQC_ENABLE;
+	if (HAL_XSPI_Init(&dev_data->hmspi.xspi) != HAL_OK) {
+		LOG_ERR("XSPI mem type MACRONIX failed");
+		return -EIO;
+	}
+
+	if (mspi_stm32_xspi_mem_ready(dev, MSPI_IO_MODE_OCTAL, cfg_rate) != 0) {
+		LOG_ERR("XSPI flash busy failed");
+		return -EIO;
+	}
+
+	if (mspi_stm32_xspi_read_cfg2reg(dev, MSPI_IO_MODE_OCTAL, cfg_rate, reg) != 0) {
+		LOG_ERR("MSPI flash config read failed");
+		return -EIO;
+	}
+
+	LOG_DBG("XSPI flash config is OCTO / %s",
+		((cfg_rate == MSPI_DATA_RATE_SINGLE) ? (char *)"STR" : (char *)"DTR"));
+
+	return 0;
+}
+
+static void mspi_stm32_xspi_isr(const struct device *dev)
+{
+	struct mspi_stm32_data *dev_data = dev->data;
+
+	HAL_XSPI_IRQHandler(&dev_data->hmspi.xspi);
+
+	k_sem_give(&dev_data->sync);
+	pm_policy_state_lock_put(PM_STATE_SUSPEND_TO_IDLE, PM_ALL_SUBSTATES);
+	(void)pm_device_runtime_put(dev);
+}
+
+#if !defined(CONFIG_SOC_SERIES_STM32H7X)
+/* weak function required for HAL compilation */
+__weak HAL_StatusTypeDef HAL_DMA_Abort_IT(DMA_HandleTypeDef *hdma)
+{
+	return HAL_OK;
+}
+
+/* weak function required for HAL compilation */
+__weak HAL_StatusTypeDef HAL_DMA_Abort(DMA_HandleTypeDef *hdma)
+{
+	return HAL_OK;
+}
+#endif /* !CONFIG_SOC_SERIES_STM32H7X */
+
+static void mspi_stm32_xspi_dma_callback(const struct device *dev, void *arg, uint32_t channel,
```

Comment:
In case the transfer is not using DMA, then this mspi_stm32_xspi_dma_callback gets a " warning: 'mspi_stm32_xspi_dma_callback' defined but not used"
In that case (no DMA) 
`#define STM32_XSPI_USE_DMA DT_NODE_HAS_PROP(STM32_XSPI_NODE, dmas)`
could flag some parts with  `#if STM32_XSPI_USE_DMA`